Added configurable max cost and update

mvez73 opened the pull request:
Added the possibility to change the maximum cost with a config.yml file and update to spigot API 1.18.2 and ProtocolLib 4.8.0

mvez73 committed:
Added the possibility to change the maximum cost with a config.yml file and update to spigot API 1.18.2 and ProtocolLib 4.8.0

Jikoo commented on pom.xml:
```diff
@@ -15,28 +15,37 @@
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
```
Guava is provided by Spigot and does not need to be included in the dependency list. Also, this plugin doesn't use it.

Jikoo commented on src/main/java/com/github/jikoo/anvilunlocker/AnvilUnlocker.java:
```diff
@@ -22,9 +22,10 @@ public void onEnable() {
 
 	@EventHandler
 	public void onInventoryOpen(InventoryOpenEvent event) {
+		int maximumCost = getConfig().getInt("maximumcost");
```
Would prefer this to be inside the if statement - that way we aren't pulling a config value every time any inventory opens. It's a small optimization for sure, but an easy one.

Jikoo commented on src/main/java/com/github/jikoo/anvilunlocker/AnvilUnlocker.java:
```diff
@@ -22,9 +22,10 @@ public void onEnable() {
 		if (event.getInventory() instanceof AnvilInventory && event.getPlayer() instanceof Player
 				&& event.getPlayer().getGameMode() != GameMode.CREATIVE) {
-			((AnvilInventory) event.getInventory()).setMaximumRepairCost(Short.MAX_VALUE);
+			((AnvilInventory) event.getInventory()).setMaximumRepairCost(maximumCost);
```
Does this actually work? I imagine we need to toggle faked gamemode post-PrepareAnvilEvent state so that we're displaying "Too Expensive!" when the result is too expensive. The result should be denied, but does the client understand that? I imagine it will create some desync issues.

Jikoo replied:
After experimenting I found that the lack of "Too Expensive!" was very confusing for an otherwise legal combination, so the post-PrepareAnvilEvent toggle is in fact necessary.

Jikoo commented on pom.xml:
```diff
@@ -15,28 +15,37 @@
+        </repository>
+        <repository>
+            <id>sonatype</id>
+            <url>https://oss.sonatype.org/content/groups/public/</url>
```
Unnecessary repo.

Jikoo commented on src/main/resources/plugin.yml:
```diff
@@ -1,6 +1,6 @@
 main: ${project.groupId}.${project.artifactId}.${project.name}
 version: ${project.version}
-api-version: "1.15"
+api-version: "1.18"
```
This forces the users of the plugin to be running 1.18 when it can still run just fine on the version it was originally written against - the packet format has not changed. I get that 1.18 is the majority market share, but I don't see a reason to break backwards compatibility in this case. This is also why I didn't bother updating dependencies.

Jikoo replied:
I guess we may as well update and go with recent Java constructs though, it'll clean up the code a touch.

Jikoo committed:
Start suggested changes

Jikoo committed:
Use new language constructs

mvez73 commented:
Guava is provided by Spigot and does not need to be included in the
dependency list. Also, this plugin doesn't use it.
Copy/paste my pom.xml. I forgot to remove it

Would prefer this to be inside the if statement - that way we aren't
pulling a config value every time any inventory opens. It's a small
optimization for sure, but an easy one.
Ok

Does this actually work? I imagine we need to toggle faked gamemode
post-PrepareAnvilEvent state so that we're displaying "Too Expensive!" when
the result is too expensive. The result should be denied, but does the
client understand that? I imagine it will create some desync issues.
Yes it works. I tested it. One thing i wanted to do in my plugin is check
if the cost is over the limit setted in the config file and not calling
your method. So it will display !Too expensive!" I didn't succeed to import
the variable from my class to yours.

+        <repository>
+            <id>sonatype</id>

+ <url>https://oss.sonatype.org/content/groups/public/</url>
Unnecessary repo.
By default the Minecraft plugin in IntelliJ add this repo. I don't know why
if it's not needed...

-api-version: "1.15"
+api-version: "1.18"

This forces the users of the plugin to be running 1.18 when it can still
run just fine on the version it was originally written against - the packet
format has not changed. I get that 1.18 is the majority market share, but I
don't see a reason to break backwards compatibility in this case. This is
also why I didn't bother updating dependencies.
I don't think it does... I'll check right after sending you this email but
i think even if you use the 1.18 API you still can use the plugin with 1.14
for example. The java version is différent i think. You have to use at
least the version in the plugin. That's why most people keep using 1.8 for
java 8

Le sam. 23 avr. 2022 à 08:40, Adam ***@***.***> a écrit :

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/resources/plugin.yml
> <https://github.com/Jikoo/AnvilUnlocker/pull/1#discussion_r856888334>:
>
> > @@ -1,6 +1,6 @@
>  name: ${project.name}
>  main: ${project.groupId}.${project.artifactId}.${project.name}
>  version: ${project.version}
> -api-version: "1.15"
> +api-version: "1.18"
>
> I guess we may as well update and go with recent Java constructs though,
> it'll clean up the code a touch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Jikoo/AnvilUnlocker/pull/1#discussion_r856888334>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM3JBWGVM7TKI5SNFTLA5ADVGPVSFANCNFSM5UD3TOCQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Jikoo committed:
Use PrepareAnvilEvent to determine necessary fake gamemode

Should ideally result in fewer desync issues and correct "Too Expensive!" display.

Jikoo committed:
Actually save config, clamp and cache value

Also set default to old default.

Jikoo commented:
Made all the changes I requested 👍

> Yes it works. I tested it. One thing i wanted to do in my plugin is check if the cost is over the limit setted in the config file and not calling your method

Specifically, if the maximumCost is set to 45 and the cost is 50 it will display "Too Expensive!"? I didn't think the creative client was capable of that - last I was aware the max cost is never communicated to the client. If that is the case we could revert the changes I added in fff63ef35f and reduce load by not sending as many packets and not using the scheduler.

Jikoo commented:
Sorry, missed some of your replies due to the editor fold, GitHub's email reply system appears to have quoted the entire end section by accident.

> > This forces the users of the plugin to be running 1.18 when it can still run just fine on the version it was originally written against - the packet format has not changed. I get that 1.18 is the majority market share, but I don't see a reason to break backwards compatibility in this case. This is also why I didn't bother updating dependencies.
> 
> I don't think it does... I'll check right after sending you this email but i think even if you use the 1.18 API you still can use the plugin with 1.14 for example. The java version is différent i think. You have to use at least the version in the plugin. That's why most people keep using 1.8 for java 8

Unfortunately, no, that's not how it works. `api-version` was introduced in 1.13, so versions before that do not support it. However, versions later than that will not load plugins that depend on a newer `api-version`. They will throw an InvalidPluginException.

Version check is done on load: https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java?until=f8b2086d60942eb2cd7ac25a2a1408cb790c222c&untilPath=src%2Fmain%2Fjava%2Forg%2Fbukkit%2Fplugin%2Fjava%2FJavaPluginLoader.java#125
Actual internal version handling is inside CraftMagicNumbers:
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java?until=86292d3e084729f53ce77af17b7b20d91754b7a5&untilPath=src%2Fmain%2Fjava%2Forg%2Fbukkit%2Fcraftbukkit%2Futil%2FCraftMagicNumbers.java#309
If you look at the constants specified above, you can see exactly which version strings are legal.

Java version is not exactly related, but certain guarantees can be made. Ex: in 1.11, Mojang released Minecraft compiled against Java 8, which meant that anyone using 1.11 and higher is guaranteed to be using over Java 8. They then upgraded to Java 16 with 1.17 and again to Java 17 with 1.18. By using `api-version: "1.18"` we guarantee it is okay to compile using Java 17, because the plugin won't enable on earlier servers anyway.

Jikoo committed:
Revert Spigot/language version bumps

Jikoo committed:
Fix "Too Expensive!" with no result

Actual solution is to not display it for no secondary input - illegal combination also yields no result, but client can figure out that case.
